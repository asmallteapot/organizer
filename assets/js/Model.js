import React from 'react'
import Events from 'ampersand-events'
import AmpersandModel from 'ampersand-model'
import AmpersandCollection from 'ampersand-collection'
import AmpersandRestCollection from 'ampersand-rest-collection'
import _ from 'lodash'
import moment from 'moment'
import slug from 'slug'

import { csrftoken } from './Django'

const DjangoConfig = () => {
  return {
    headers: {
      'X-CSRFToken': csrftoken
    }
  }
}

export const DjangoModel = AmpersandModel.extend({
  ajaxConfig: DjangoConfig,
  props: {
    id: 'number'
  },
  dataTypes: {
    moment: {
      set(v) {
        return {type: 'moment', val: moment(v)};
      },
      default() {
        return moment();
      }
    }
  },
  derived: {
    url: {
      deps: ['id'],
      fn() {
        return this.getId() ? (this.urlRoot + this.getId() + '/') : this.urlRoot;
      }
    }
  }
})

export const DjangoCollection = AmpersandRestCollection.extend({
  ajaxConfig: DjangoConfig,
  parse(response) {
    if (response.results) {
      return response.results;
    } else {
      return response;
    }
  }
})

export const Activist = DjangoModel.extend({
  urlRoot: '/api/activists/',
  props: {
    address: 'string',
    created: 'string',
    email: 'string',
    name: 'string',
    rank: 'number',
  }
});

export const Signup = DjangoModel.extend({
  urlRoot: '/api/signups/',
  props: {
    state: 'string',
    action: 'string',
  },
  toJSON() {
    return {...this.serialize(), ...{activist: this.activist.url}};
  },
  children: {
    activist: Activist
  }
});

export const SignupCollection = DjangoCollection.extend({
  url: '/api/signups/',
  model: Signup,
})

export const ActivistCollection = DjangoCollection.extend({
  url: '/api/activists/',
  model: Activist
})

export const Action = DjangoModel.extend({
  urlRoot: '/api/actions/',
  props: {
    name: ['string', true, () => "Untitled"],
    date: ['moment', true, () => moment()],
    fields: ['array', false, () => []],
    forms: ['array', false, () => []]
  },
  derived: {
    slug: {
      deps: ['name'],
      type: 'string',
      fn() {
        return slug(this.name);
      }
    }
  },
  collections: {
    signups: SignupCollection,
    //forms: FormCollection,
    //fields: FieldCollection
  }
});

export const ActionCollection = DjangoCollection.extend({
  url: '/api/actions/',
  model: Action
})

export const Form = DjangoModel.extend({
  urlRoot: '/api/forms/',
  props: {
    active: 'boolean',
    description: 'string',
    next_state: 'string',
    title: 'string'
  },
  collections: {
    //fields: FieldCollection
  },
  children: {
    action: Action
  }
});

export const FormCollection = DjangoCollection.extend({
  url: '/api/forms/',
  model: Form
});

export const SubmissionField = AmpersandModel.extend({
  props: {
    id: 'number',
    value: 'string'
  }
});

export const SubmissionFieldCollection = AmpersandCollection.extend({
  model: SubmissionField
});

export const Submission = DjangoModel.extend({
  props: {
    name: 'string',
    email: 'string',
    address: 'string',
    form: Form,
  },
  collections: {
    fields: SubmissionFieldCollection
  },
  toJSON() {
    const serialized = this.serialize('props');
    var fieldValues = {}
    _.each(this.fields, field => {
      fieldValues['input_'+field.id] = field.value;
    })
    return {...serialized, ...fieldValues};
  },
  derived: {
    url: {
      deps: ['*'],
      fn() {
        return '/api/forms/'+this.form.id+'/submit_response/'
      }
    }
  }
});

export function bindToState(self, state, propMap) {
  _.each(_.keys(propMap), propName => {
    state.on('change:'+propName, () => {
      var nextState = {};
      nextState[propName] = state[propName];
      self.setState(nextState);
    })
  })
}

export function bindToCollection(self, collection, key) {
  collection.on('add remove change', () => {
    var nextState = {};
    nextState[key] = collection.models;
    self.setState(nextState);
  });
}

export function withState(Component) {
  return class StateBinding extends React.Component {
    constructor(props) {
      super(props);
      Events.createEmitter(this);
      this.onUpdate = _.debounce(this.onUpdate.bind(this), 100);
      this.state = {updateHash: 0};
      this.mounted = false;
    }

    componentDidMount() {
      this.mounted = true;
      this.rebindProps({}, this.props);
    }

    componentWillUnmount() {
      this.mounted = false;
      this.stopListening();
    }

    onUpdate() {
      if (this.mounted) {
        this.setState({updateHash: this.state.updateHash+1});
      }
    }

    rebindProps(oldProps, nextProps) {
      const nextStates    = _.filter(_.values(nextProps), p => p.isState);
      const currentStates = _.filter(_.values(oldProps), p => p.isState);
      const removedStates = _.difference(currentStates, nextStates);
      const addedStates   = _.difference(nextStates, currentStates);

      _.each(removedStates, this.stopListening);
      _.each(addedStates,   s => this.listenTo(s, 'change', this.onUpdate));
    }

    componentWillReceiveProps(nextProps) {
      this.rebindProps(this.props, nextProps);
    }

    render() {
      return <Component {...this.props} />
    }
  }
}
