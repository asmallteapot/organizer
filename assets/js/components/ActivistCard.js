import { DragSource } from 'react-dnd'
import React from 'react'
import Gravatar from 'react-gravatar'
import { ContextMenuTrigger, ContextMenu, MenuItem } from 'react-contextmenu'
import _ from 'lodash'
import { Checkbox } from 'react-form'

class ActivistCardBase extends React.Component {
  render() {
    const classes = "card " +
                    (this.props.dragging ? "dragging " : "");
    const draggable = this.props.connectDragSource(
      <div className={classes} key={this.props.activist.cid}>
        <ContextMenuTrigger id={this.props.activist.cid}>
          <div className="right">
            <div className={"rank rank-" + this.props.activist.rank}>{this.props.activist.rank}</div>
            <ContextMenuTrigger showOnClick={true} holdToDisplay={-1} id={this.props.activist.cid}>
              <button className="options"><i className="fa fa-ellipsis-v" /></button>
            </ContextMenuTrigger>
          </div>
          <div className="name">
            <Gravatar size={24} email={this.props.activist.email} />
            {this.props.activist.name}
          </div>
          <div className="email">{this.props.activist.email}</div>
        </ContextMenuTrigger>
      </div>
    )
    return (
      <div className="card-wrap">
        {draggable}
        <ContextMenu id={this.props.activist.cid}>
        </ContextMenu>
      </div>
    )
  }
}

const dragSpec = {
  beginDrag(props, monitor, component) {
    props.onDragging && props.onDragging(true);
    return {activist: props.activist}
  },
}

const collect = (connect, monitor) => ({
  connectDragSource: connect.dragSource(),
  dragging: monitor.isDragging()
})

const ActivistDragSource = DragSource("activist", dragSpec, collect);

const ActivistCard = ActivistDragSource(ActivistCardBase);
export default ActivistCard;
