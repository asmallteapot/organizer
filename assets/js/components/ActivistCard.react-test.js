import React from 'react'
import { mount } from 'enzyme'
import ActivistCard from './ActivistCard'
import { Signup } from '../Model'
import { DragDropContext } from 'react-dnd'
import TestBackend from 'react-dnd-test-backend'

const WrappedCard = DragDropContext(TestBackend)(ActivistCard)

test('Smoke test', () => {
  const signup = new Signup();
  const component = mount(<WrappedCard signup={signup} />)
  expect(component.html()).toMatchSnapshot()
});
