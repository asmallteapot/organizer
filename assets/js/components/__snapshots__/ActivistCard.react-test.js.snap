// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Smoke test 1`] = `
"<div class=\\"activist-card-wrap\\"><div><div class=\\"activist-card \\" style=\\"color: rgba(0, 0, 0, 0.87); background-color: rgb(255, 255, 255); box-sizing: border-box; font-family: Roboto, sans-serif; box-shadow: 0 1px 6px rgba(0, 0, 0, 0.12),
         0 1px 4px rgba(0, 0, 0, 0.12); border-radius: 2px; z-index: 1;\\"><div style=\\"padding-bottom: 0px;\\"><div style=\\"padding: 16px; font-weight: 500; box-sizing: border-box; position: relative; white-space: nowrap;\\"><div size=\\"40\\" class=\\"rank-undefined\\" style=\\"color: rgb(255, 255, 255); background-color: rgb(188, 188, 188); display: inline-flex; font-size: 20px; border-radius: 50%; height: 40px; width: 40px; margin-right: 16px;\\"></div><div style=\\"display: inline-block; vertical-align: top; white-space: normal; padding-right: 90px;\\"><span style=\\"color: rgba(0, 0, 0, 0.87); display: block; font-size: 15px;\\"></span><span style=\\"color: rgba(0, 0, 0, 0.54); display: block; font-size: 14px;\\"></span></div></div></div></div></div></div>"
`;
